iothub/device/samples/convention-based-samples/TemperatureController/TemperatureControllerSample.cs

```diff
@@ -0,0 +1,372 @@
+        private async Task HandlePropertyUpdatesAsync(ClientPropertyCollection writableProperties, object userContext)
+        {
+            foreach (KeyValuePair<string, object> writableProperty in writableProperties)
+            {
+                // The dispatcher key will be either a root-level property name or a component name.
+                switch (writableProperty.Key)
+                {
+                    case Thermostat1:
+                    case Thermostat2:
+                        const string targetTemperatureProperty = "targetTemperature";
+                        if (writableProperties.TryGetValue(writableProperty.Key, targetTemperatureProperty, out double targetTemperatureRequested))
+                        {
+                            await HandleTargetTemperatureUpdateRequestAsync(writableProperty.Key, targetTemperatureRequested, writableProperties.Version, userContext);
+                            break;
+                        }
+                        else
+                        {
+                            _logger.LogWarning($"Property: Received an unrecognized property update from service for component {writableProperty.Key}:" +
+                                        $"\n[ {writableProperty.Value} ].");
+                            break;
+                        }
+
+                    default:
+                        _logger.LogWarning($"Property: Received an unrecognized property update from service:" +
+                                        $"\n[ {writableProperty.Key}: {writableProperty.Value} ].");
+                        break;
+                }
+            }
+        }
+
+        private async Task HandleTargetTemperatureUpdateRequestAsync(string componentName, double targetTemperature, long version, object userContext)
+        {
+            const string targetTemperatureProperty = "targetTemperature";
+            _logger.LogDebug($"Property: Received - component=\"{componentName}\", [ \"{targetTemperatureProperty}\": {targetTemperature}°C ].");
+
+            _temperature[componentName] = targetTemperature;
+            var reportedProperty = new ClientPropertyCollection();
+            reportedProperty.Add(componentName, targetTemperatureProperty, _temperature[componentName], StatusCodes.Accepted, version, "Successfully updated target temperature.");
+
+            ClientPropertiesUpdateResponse updateResponse = await _deviceClient.UpdateClientPropertiesAsync(reportedProperty);
+
+            _logger.LogDebug($"Property: Update - component=\"{componentName}\", {reportedProperty.GetSerializedString()} is {nameof(StatusCodes.OK)} " +
+                $"with a version of {updateResponse.Version}.");
+        }
+
+        // The callback to handle command invocation requests.
+        private Task<CommandResponse> HandleCommandsAsync(CommandRequest commandRequest, object userContext)
+        {
+            switch (commandRequest.ComponentName)
```

Comment:
so this will always need to be nested switches => switch on component name, and per-component switch on command name. Additionally, for the default case (component name = null) switch on the command name.

Adding an if-else for the component name will also work the same way, right -> if (component-name == expected value1) else if (component-name == expected value) else if (component name == null); ans each of the if blocks again having either if-else/ switch cases.

Reply:
even though this is a "command" callback, the only way to highlight the command name before the component name is to have the componentname == null case first.

Reply:
I agree there needs to be some kind of pivot on the component name. But the switch case within default of one switch case for root level felt out of place. There could be if - else for the component name and separate switch cases for each. If you had switch on the command name, it would still work but you'll have to check component name and throw if it was not present. I was suggesting more of this format:

// Handle Commands defined on root level
If(commandRequest.ComponentName == null)
{
swtich(commandName)
}
// Handle commands defined on component level
else
{
switch (componentName):
.....
}

Ofcourse you can also have  if (component-name == expected value1) else if (component-name == expected value) else if (component name == null);  as you suggested. But at least separating out root level and component level seems important to me in this example.

Reply:
We can keep going back and forth on this as it is an implementation style. But I would add more comments to what you already have to add clarity if that is the direction you take.

Reply:
sure, I'll add some comments to describe the approach; leaving it as-is does make it a bit ambiguous.